### PR TITLE
Error out when struct size is bigger than int.MaxValue

### LINKED ISSUE #97412: The type loader has silent integer overflows
From:  https://github.com/dotnet/runtime/issues/95193#issuecomment-1885347542

These silent integer overflows are present in handling of LayoutKind.Sequential as well. For example, this:

```cs
Console.WriteLine(sizeof(X));

struct X
{
   byte x;
   BigArray a;
}

[StructLayout(LayoutKind.Sequential, Size = int.MaxValue)] 
struct BigArray
{
}
```

prints -2147483648.

### PR DESCRIPTION
**Pull request by fanyang-mono**
Fixes: #97412

`sizeof()` in C# returns a signed integer. This PR is to adjust the runtime to match with that.

Added a check for struct size and a test.

### CODE REVIEW

**Comment by jkotas** [comment]
Do we need a test?

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3656]: I would use the existing `IDS_CLASSLOAD_FIELDTOOLARGE` string for this.

File: src/coreclr/vm/classlayoutinfo.cpp
Comment (jkotas) [inline src/coreclr/vm/classlayoutinfo.cpp:151]: Does this also cover the case where we have a struct with many fields (each field is less than `FIELD_OFFSET_LAST_REAL_OFFSET` in size) and the size of all fields together overflows Int32?

Comment (fanyang-mono) [reply src/coreclr/vm/classlayoutinfo.cpp:151]: No, I could do that. the thing is `FIELD_OFFSET_LAST_REAL_OFFSET` is smaller than `int.MaxValue`. So this would error out too.
```c#
[StructLayout(LayoutKind.Sequential, Size = int.MaxValue)] 
struct BigArray
{
}
```
Is that expected? I wasn't sure.

Comment (jkotas) [reply src/coreclr/vm/classlayoutinfo.cpp:151]: This should continue to work. There is no reason to break that.

Note that `struct X { BigArray a; byte x; }` fails to load today (that's correct). It is only `struct X { byte x; BigArray a; }` that loads successfully. It suggests that we do have the right checks that prevent overflows of field offsets in place, but we are missing a check that would catch overflow of the total size.

Comment (fanyang-mono) [reply src/coreclr/vm/classlayoutinfo.cpp:151]: That's interesting. I will look into it.

Comment (fanyang-mono) [reply src/coreclr/vm/classlayoutinfo.cpp:151]: @jkotas Should we allow this or not?
```
[StructLayout(LayoutKind.Sequential, Size = int.MaxValue)] 
        struct BigArray
        {
        }
```

**Comment by fanyang-mono** [comment]
@jkotas This PR is ready to review again.

**Comment by MichalStrehovsky** [comment]
I've pushed out a change that makes further progress in the handling within the managed type system, but the test probably still won't compile because `BigArray` with size of `int.MaxValue` gets its size bumped past `int.MaxValue` through alignment after we're done with layout (the final size needs to be divisible by pointer size and `int.MaxValue` isn't.

CoreCLR is fine with this overflowing but the managed type system isn't. I'm not sure if the overflow is an actual _problem_ but fixing it doesn't look exactly trivial.

File: src/coreclr/vm/methodtablebuilder.cpp
Comment (jkotas) [inline src/coreclr/vm/methodtablebuilder.cpp:4317]: It is suspicious that this does not take alignment into account and that the check for overflow is not right here.

I think the fix should be structured more like the fix that Michal pushed for the managed type system. Always check for overflow right before the (existing) calculation or assignment would actually overflow. It should not require adding new ways to compute sizes.

File: src/coreclr/tools/Common/TypeSystem/Common/LayoutInt.cs
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/Common/LayoutInt.cs:100]: Nit: My preferred way to do overflow checks is do the calculation with higher precision and then verify that there is no information lost by truncation to the original precision:
```
long result = (long)left._value + (long)right._value;
if ((int)result != result)
   throw
```
This is universal pattern that works for all operations.

File: src/coreclr/vm/methodtablebuilder.cpp
Comment (fanyang-mono) [reply src/coreclr/vm/methodtablebuilder.cpp]: Inside `InitializeFieldDescs`, alignment didn't seem to be taken into consideration. I looked around and found the following place, where it is the size of the Sequential struct. So I added a check there to see it overflows `int`.

https://github.com/dotnet/runtime/blob/a86987ccab917433d065fe5dc8870fc261f79d14/src/coreclr/vm/methodtablebuilder.cpp#L1765

**Comment by fanyang-mono** [comment]
@MichalStrehovsky Is there a way to stop the new test from being compiled by NativeAOT?

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky Is there a way to stop the new test from being compiled by NativeAOT?

Only by placing `<NativeAotIncompatible>true</NativeAotIncomplatible>` in the project file. It would be good not to disable all existing ManagedSequential type loader testing just because the new test doesn't work (i.e. put it in a separate project file).

Note that this is the managed type system that is shared with crossgen2. Might want to run crossgen2 outerloop testing to see whether this also needs to be removed from precompilation there.

The managed type system doesn't handle this test because the size of the `BigArray` struct is actually over 2 GB when boxed (this is also true on JIT-based CoreCLR). If a 2 GB overflow is problematic, is the 2 GB overflow of that one not problematic?

**Comment by fanyang-mono** [comment]
@MichalStrehovsky Based on Jan's comment from earlier. Loading `BigArray` itself should continue to work. Why do we need to align `BigArray` to pointer size? If that's needed, `sizeof(BigArray)` would return a bigger value than `int.MaxValue`. Is that correct?

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky Based on Jan's comment from earlier. Loading `BigArray` itself should continue to work. Why do we need to align `BigArray` to pointer size? If that's needed, `sizeof(BigArray)` would return a bigger value than `int.MaxValue`. Is that correct?

The size of it gets aligned when boxing it, because it's 8 bytes larger (MethodTable pointer) and aligned. On CoreCLR today, doing:

```csharp
BigArray.Instance = new BigArray();

long before = GC.GetAllocatedBytesForCurrentThread();
BigArray.Instance = new BigArray();
Console.WriteLine(GC.GetAllocatedBytesForCurrentThread() - before);

[StructLayout(LayoutKind.Sequential, Size = int.MaxValue)]
struct BigArray
{
    public static object Instance;
}
```

Will print 2,147,483,664, which is larger than int.MaxValue.

**Comment by MichalStrehovsky** [comment]
> The size of it gets aligned when boxing it, because it's 8 bytes larger (MethodTable pointer) and aligned. On CoreCLR today, doing:

We chatted with Fan on Teams about this and apparently the program I posted only works on a Release build (also _without_ debugger attached) and throws InvalidProgram otherwise. On .NET 8 CoreCLR-JIT. Curious.

**Comment by MichalStrehovsky** [comment]
Actually, isn't the negative size in https://github.com/dotnet/runtime/issues/97412 just a "display artifact"? ECMA-335 defines sizeof as returning an unsigned int32, but the C# compiler treats it as signed. So it looks like it overflowed, but that's just because the C# compiler incorrectly retyped the number.

**Comment by fanyang-mono** [comment]
Based on the discussion here: https://github.com/dotnet/runtime/issues/95193#issuecomment-1906592092, the conclusion was that the size of struct shouldn't be bigger than `FIELD_OFFSET_LAST_REAL_OFFSET`.

**Comment by fanyang-mono** [comment]
@MichalStrehovsky You mentioned earlier that when boxing `BigArray`, it got aligned to pointer size. That shouldn't affect the return value of `sizeof(BigArray)`, meaning the alignment due to boxing shouldn't be taken into consideration. That's implementation details related to boxing operation, not sizeof.

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky You mentioned earlier that when boxing `BigArray`, it got aligned to pointer size. That shouldn't affect the return value of `sizeof(BigArray)`, meaning the alignment due to boxing shouldn't be taken into consideration. That's implementation details related to boxing operation, not sizeof.

In general i don't understand the reason why a int32 limit is needed here. I assume this spot also limits size of sequential layout classes. If it affects classes, why a class cannot cross this limit, why a boxed struct can. And if it doesn't affect classes, why not?

The motivation with C# sizeof feels a bit bogus because I believe that's a C# bug. I can imagine we have problens with large enough classes/structs but I'm not very confident max int32 is the limit and LAST_REAL_OFFSET has a good chance to be breaking while at the same time not being the actual problem threshold.

**Comment by fanyang-mono** [comment]
@MichalStrehovsky Please refer to https://github.com/dotnet/runtime/issues/95193#issuecomment-1906592092 and the discussion above it. My takeaway after reading that discussion is that it seems to be meaningless to allow a struct of a super big size, which may cause issues later on during runtime. Let's agree on a large enough number. That is `FIELD_OFFSET_LAST_REAL_OFFSET`.

**Comment by fanyang-mono** [comment]
We should align NativeAOT/crossgen2 with the same behavior as CoreCLR before merging this PR.

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky Please refer to [#95193 (comment)](https://github.com/dotnet/runtime/issues/95193#issuecomment-1906592092) and the discussion above it. My takeaway after reading that discussion is that it seems to be meaningless to allow a struct of a super big size, which may cause issues later on during runtime. Let's agree on a large enough number. That is `FIELD_OFFSET_LAST_REAL_OFFSET`.

That discussion is about size of InlineArray which was a new feature in .NET 8. What this PR is doing is adding limit to something that didn't have this limit for 22+ years (it's not a new thing like InlineArray). It is a breaking change: we better have a good justification for customers who may be broken by it. `LAST_REAL_OFFSET` feels pretty low. Real world users might be broken - this has been possible for 22 years after all and we don't know if type sizes above this limit are problematic at all.

I already wrote my opinion in the InlineArray issue that the limit should be where we have implementation limitations that may cause it to be a problem. Instead of doing that, an arbitrary limit was chosen (`LAST_REAL_OFFSET` limit is based on VMs need to steal 5 bits of field offset during type loading - which doesn't apply to the situation we're discussing and it didn't apply to InlineArray either).

At least one limitation seems to be that the C# language decided maximum size of a type is 2 GB so we could as a courtesy not allow that. `sizeof` returning a negative number seems bad enough that it _may_ justify a breaking change.

I don't think anyone has an understanding where the other implementation limitations are. Native AOT and ReadyToRun probably have a limitation around _aligned size of the type_ being over 2 GB. It's not clear if there are other limitations. Since nobody investigated the implementation limitations, what is giving us the confidence that `FIELD_OFFSET_LAST_REAL_OFFSET` is better than 2 GB? What is giving us confidence things below `FIELD_OFFSET_LAST_REAL_OFFSET` will not overflow?

That said, people make changes in this area despite my contrary opinions, that's fine, I'm not the one ultimately responsible. I can help mirroring whatever approach is decided into the managed type system and can sign off on managed type system changes. I don't review CoreCLR VM changes since I'm not expert so I don't need to be convinced on this.

**Comment by fanyang-mono** [comment]
@jkotas @VSadov Do you agree with the solution here?

**Comment by jkotas** [comment]
> @jkotas @VSadov Do you agree with the solution here?

I think that the current version of the PR is unnecessary breaking change. It is rejecting a lot more than just what's required to fix the overflow.

**Comment by fanyang-mono** [comment]
@jkotas If choose `int.MaxValue` or anything number bigger than that. From the C# user point of view, it means somehow we allow the last field of a struct to be very big, but we don't allow that for other fields. This is a very strange behavior to me.

**Comment by jkotas** [comment]
The CoreCLR JIT and type loader have number of internal implementation limits, maximum field offset is one of them. I agree that it can be produce non-intuitive observable behaviors if one does not know what those limits are exactly. I am not worried about it too much. These non-intuitive behaviors are only going be hit by the few folks doing low-level tricks with extremely large structs and they should be capable to figure it out.

**Comment by fanyang-mono** [comment]
@jkotas There is another customer opened issue which complains about this inconsistent behavior. See https://github.com/dotnet/runtime/issues/105064 for more details. If we are fine with these kind of customer complains, then I could revert the solution back to checking if the size of the struct is smaller than `int.MaxValue` to be compliant with roslyn/C#.

**Comment by jkotas** [comment]
> See https://github.com/dotnet/runtime/issues/105064 for more details.

#105064 can be improved by providing a better error message (field offset too big) instead of the generic type load exception error.

My primary concern with current version of the change is that it is a breaking change. It breaks code that worked fine for last 20 years and that likely exists out there. It would need breaking change notice with a justification, etc. 

If we really wanted to reduce the inconsistencies, it would be better to do that by relaxing the type loader internal limits, without introducing breaking changes.

I think this PR should be just about fixing the `sizeof(...)` overflow problem in C#. Checking for `int.MaxValue` should be sufficient for that.

**Comment by fanyang-mono** [comment]
@MichalStrehovsky As discussed above, please mirror this behavior of CoreCLR for NativeAOT. It is currently failing.

File: src/mono/mono/metadata/class-init.c
Comment (lambdageek) [inline src/mono/mono/metadata/class-init.c:2541]: nit: 
```suggestion
	/*valuetypes can not be size 0 or bigger than 2gb. */
```

File: src/coreclr/vm/methodtablebuilder.cpp
Comment (fanyang-mono) [reply src/coreclr/vm/methodtablebuilder.cpp:1767]: For field offset, we check for both too. See
https://github.com/dotnet/runtime/blob/6931b3b729ec050f935f958122d3cc66bb46fdcf/src/coreclr/vm/methodtablebuilder.cpp#L4354-L4357

I don't have strong opinion for doing it as it is now or limiting it to valuetypes only. @jkotas Any thoughts?

Comment (jkotas) [reply src/coreclr/vm/methodtablebuilder.cpp:1767]: I think Michal's comment was trying to point out that the following program still prints `-2147483648` even with this fix:

```
using System.Runtime.InteropServices;

unsafe {
    Console.WriteLine(sizeof(X));
}

[StructLayout(LayoutKind.Explicit)] 
struct X
{
    [FieldOffset(0)]
    bool x;
    [FieldOffset(1)]
    BigArray a;
}

[StructLayout(LayoutKind.Sequential, Size = int.MaxValue)]
struct BigArray
{
}
```

Comment (jkotas) [reply src/coreclr/vm/methodtablebuilder.cpp:1767]: You want this check to apply to all structs, not just to `IsBlittable() || IsManagedSequential()` structs.

Comment (jkotas) [reply src/coreclr/vm/methodtablebuilder.cpp:1767]: > For explicit struct, I see IsBlittable() returns YES

I do not think that it does for the explicit struct in my example.

File: src/tests/Loader/classloader/SequentialLayout/ManagedSequential/LargeStructSize_CoreCLR.cs
Comment (jkotas) [inline src/tests/Loader/classloader/SequentialLayout/ManagedSequential/LargeStructSize_CoreCLR.cs:22]: This test is good, but you should also add one with `bool` that I have shared in my comment. I expect that the test case with `bool` is still going to fail with the current version of the change.

Comment (fanyang-mono) [reply src/tests/Loader/classloader/SequentialLayout/ManagedSequential/LargeStructSize_CoreCLR.cs:22]: Fixed and added a test.

File: src/coreclr/vm/methodtablebuilder.cpp
Comment (fanyang-mono) [reply src/coreclr/vm/methodtablebuilder.cpp]: Fixed and added a test.

File: src/tests/issues.targets
Comment (jkotas) [inline src/tests/issues.targets:31]: Would it be better to use `SkipOnCoreClr` attribute instead?

Comment (fanyang-mono) [reply src/tests/issues.targets:31]: I tried and hit this bug: https://github.com/dotnet/runtime/issues/105964

**Review by jkotas — APPROVED** [review]
Thanks

**Comment by fanyang-mono** [comment]
@lambdageek May I get a sign-off for Mono side of change?

**Review by lambdageek — APPROVED** [review]

**Comment by fanyang-mono** [comment]
@MichalStrehovsky CoreCLR and Mono side of work has been completed and approved. The remaining piece of work is related to NativeAOT/Crossgen2. I am going to transfer the ownership of this PR to you, since you will be working on that.

Additionally, this could probably be merged for .NET9, as I don't think that this is a breaking change. @jkotas Please correct me, if I were wrong.

**Comment by MichalStrehovsky** [comment]
> The remaining piece of work is related to NativeAOT/Crossgen2. I am going to transfer the ownership of this PR to you, since you will be working on that.

I'll not have time to look at this any time soon. Might be better to just block the test on an issue and not block this PR.

> Additionally, this could probably be merged for .NET9, as I don't think that this is a breaking change.

Throwing an exception in a situation that didn't throw before is [the definition](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/breaking-change-rules.md#exceptions) of a breaking change.

**Comment by fanyang-mono** [comment]
Discussed with @jkotas offline. He suggested to wait for .NET10 to merge it, together with the NativeAOT change, when it is ready.

There is no rush. @MichalStrehovsky

**Comment by fanyang-mono** [comment]
I plan to merge this PR when main is .NET10. The managed type system part used by NativeAOT and crossgen can be worked on later during .NET10 release.

**Comment by fanyang-mono** [comment]
CI failures are unrelated to this PR.

**Pull request merged**
